### PR TITLE
Add Listener Access Logs when ResponseFlag is set

### PR DESCRIPTION
**Pull request by gargnupur**
Ref: https://github.com/istio/istio/issues/26851

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[X ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by gargnupur** [comment]
@howardjohn , @kyessenov , @mandarjog : Currently, I have enabled them for any response flag set... should we restrict it to specific flags like "NR". Getting these logs for health check or ratelimiting errors might not be useful..

**Comment by mandarjog** [comment]
@gargnupur this needs a release note.

**Review by kyessenov — COMMENTED** [review]
Are there any tests that exercise stdout access log here? It might be worth checking functional behavior, not just protos.

**Comment by gargnupur** [comment]
> Are there any tests that exercise stdout access log here? It might be worth checking functional behavior, not just protos.

I don't think so.. but can try to write one.. I did validate it manually (by removing the filter)...

**Comment by mandarjog** [comment]
We should have examples of how and when to use this feature.

**Comment by gargnupur** [comment]
> We should have examples of how and when to use this feature.

We can add it in documentation over here : https://istio.io/latest/docs/tasks/observability/logs/

File: pilot/pkg/networking/core/v1alpha3/accesslog.go [hunk 1]
```diff
@@ -100,16 +102,22 @@ type AccessLogBuilder struct {
 	tcpGrpcAccessLog *accesslog.AccessLog
 	// httpGrpcAccessLog is used when access log service is enabled in mesh config.
 	httpGrpcAccessLog *accesslog.AccessLog
+	// tcpListenerGrpcAccessLog is used when access log service is enabled in mesh config.
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/accesslog.go:105]: nit: comment and variable name do not match.

File: pilot/pkg/networking/core/v1alpha3/accesslog.go [hunk 2]
```diff
@@ -100,16 +102,22 @@ type AccessLogBuilder struct {
...
+	tcpGrpcListenerAccessLog *accesslog.AccessLog
 
 	// file accessLog which is cached and reset on MeshConfig change.
 	mutex         sync.RWMutex
 	fileAccessLog *accesslog.AccessLog
+	// listener file accessLog which is cached and reset on MeshConfig change.
+	listenerMutex         sync.RWMutex
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/accesslog.go:112]: should we just used the `mutex`?

**Review by howardjohn — COMMENTED** [review]
looks good

File: pilot/pkg/networking/core/v1alpha3/accesslog.go (hunk 1 above)
Comment (gargnupur) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go:105]: Thanks..Fixed

File: pilot/pkg/networking/core/v1alpha3/accesslog.go (hunk 2 above)
Comment (gargnupur) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go:112]: Done..

**Review by howardjohn — APPROVED** [review]

**Comment by gargnupur** [comment]
@howardjohn , @kyessenov , @mandarjog : I verified that there are no duplicate logs by running rate limiting test(has RL and RLSE response flags) and trust domain tests.

For trust domain test, I ran the scenario of `verify(t, client, "bar", httpMTLS, scheme.HTTPS, false)` => requests from trust domain "bar" are denied.  I see following listener logs:
[2020-10-16T13:21:53.509Z] "- - -" 0 - "-" 2653 3521 11 - "-" "-" "-" "-" "10.15.253.199:8091" PassthroughCluster 10.12.0.153:44278 10.15.253.199:8091 10.12.0.153:44276 server.trust-domain-validation-1-10212.svc.cluster.local -
It's weird that ResponseFlag is not set above...

**Comment by gargnupur** [comment]
@howardjohn , @kyessenov: Summarizing the offline chat
The format of the log for denied request because of untrusted domain is as follows:
[2020-10-16T21:56:04.245Z] "- - -" 0 UF,URX "-" 0 0 1028 - "-" "-" "-" "-" "10.15.240.12:8091" PassthroughCluster - 10.15.240.12:8091 10.12.1.76:45248 - - -
But we get duplicate logs with TCP as follows:
[2020-10-16T21:56:04.245Z] "- - -" 0 UF,URX "-" 0 0 1028 - "-" "-" "-" "-" "10.15.240.12:8091" PassthroughCluster - 10.15.240.12:8091 10.12.1.76:45248 - -

Thus, to deal with duplicate logs with TCP we are doing follows:
1) Change to just "NR" response flag
2) Default on, but add a config to disable if someone doesn't want this -> https://github.com/istio/api/pull/1697

/cc @mandarjog

**Comment by gargnupur** [comment]
@howardjohn , @kyessenov : Added config to disable the logs too. PTAL at it, I will remove the merge-hold after that..

**Review by howardjohn — COMMENTED** [review]
LGTM

**Pull request merged**